### PR TITLE
OCPCLOUD-1247: Set correct kubelet hostname for AWS platform

### PR DESCRIPTION
**Pull request by Fedosin**
It allows to define type for the platform where the cluster is deployed.

If platform type is AWS, then we explicitly override kubelet hostname with the value from Metadata.

It is required for deploying clusters on Windows machines with Cloud Controller Manager support, where hostnames must be the same as node names.

Related links:
https://issues.redhat.com/browse/OCPCLOUD-1247
https://bugzilla.redhat.com/show_bug.cgi?id=1995894

### CODE REVIEW

File: pkg/bootstrapper/bootstrapper.go
Comment (JoelSpeed) [inline pkg/bootstrapper/bootstrapper.go:411]: It's not usual to add a new line to an error string like this, what's the intention here?

Also, is there a particular purpose to wrapping the error? Are you likely to unwrap later? If not, just use `%v` instead, it's more idiomatic

File: pkg/bootstrapper/bootstrapper.go
Comment (JoelSpeed) [inline pkg/bootstrapper/bootstrapper.go:576]: This switch could grow over time, so as not to distract from the purpose of this function, why not make this a helper?

```
func getKubeletHostname(platform configv1.PlatformType) (string, error) {
  switch platform {
  case configv1.AWSPlatformType:
    return getAWSMetadataHostname()
  default:
    return "", nil
  }
}
```

File: pkg/bootstrapper/hostname.go
Comment (JoelSpeed) [inline pkg/bootstrapper/hostname.go:12]: Do we want to have the AWS specific functions within either a file with AWS in the name or a separate provider specific package?

File: pkg/bootstrapper/hostname.go
Comment (JoelSpeed) [inline pkg/bootstrapper/hostname.go:31]: Would recommend not wrapping the errors here and removing the the new lines as well

File: pkg/bootstrapper/bootstrapper.go
Comment (sebsoto) [reply pkg/bootstrapper/bootstrapper.go:576]: Agreed, I'd also like to see this moved to a separate package

File: pkg/bootstrapper/hostname.go
Comment (sebsoto) [reply pkg/bootstrapper/hostname.go:12]: Another package for all providers would be preferred with a separate AWS file.

File: test/e2e/bootstrapper_test.go
Comment (sebsoto) [inline test/e2e/bootstrapper_test.go:65]: The provider should be discovered as part of the test and then passed in to all invocations of `NewWinNodeBootstrapper`, excluding the uninstall kubelet tests, as arguments are ignored there.

A test should be added to check that the `hostname-override` argument is set.

Do we need to check the value? What is the expected behavior if the value is incorrect?

File: cmd/bootstrapper/initialize_kubelet.go
Comment (sebsoto) [inline cmd/bootstrapper/initialize_kubelet.go:62]: Can you add example values, such as `aws`?

File: pkg/bootstrapper/bootstrapper.go
Comment (sebsoto) [inline pkg/bootstrapper/bootstrapper.go:571]: lets make this  `aws` instead of `AWS`, as this is a flag being passed in, and all lowercase is more commonly used

Comment (sebsoto) [reply pkg/bootstrapper/bootstrapper.go:571]: This could potentially be made a constant in the new package you will introduce.

**Review by sebsoto — CHANGES_REQUESTED** [review]

File: pkg/bootstrapper/bootstrapper.go
Comment (Fedosin) [reply pkg/bootstrapper/bootstrapper.go:576]: moved to a new package "cloud"

File: pkg/bootstrapper/hostname.go
Comment (Fedosin) [reply pkg/bootstrapper/hostname.go:12]: done

File: pkg/bootstrapper/hostname.go
Comment (Fedosin) [reply pkg/bootstrapper/hostname.go:31]: done

File: test/e2e/bootstrapper_test.go
Comment (Fedosin) [reply test/e2e/bootstrapper_test.go:65]: I put "azure" platform as the parameter value there at the moment, in this case wcmb doesn't override the hostname. In case of aws, it will send requests to AWS metadata service, which is hard to emulate in these tests.

File: pkg/bootstrapper/cloud/cloud.go
Comment (JoelSpeed) [inline pkg/bootstrapper/cloud/cloud.go]: Nit: I'd probably call this `GetKubeletHostnameOverride` so that it looks more like the kubelet flag `hostname-override`, but I'll defer to WMCO folks as to whether they think that's a required change

File: pkg/bootstrapper/cloud/aws.go
Comment (JoelSpeed) [inline pkg/bootstrapper/cloud/aws.go]: Is there a reference somewhere for why we are using this value? Perhaps a link to the old cloud provider code or something? It would be good to link the original source if possible

**Comment by JoelSpeed** [comment]
Having watched the recording from the monday meeting, I wanted to add some notes about this to alleviate some concerns about supportability that were raised and some clarifications on technical details:
- With in-tree, Kubelet asks the cloud what name it should use for the node. In the AWS implementation, this uses the [PrivateDNSName](https://github.com/kubernetes/legacy-cloud-providers/blob/c0697affa6e0a5445bbde000c89e4d928d44969d/aws/aws.go#L4926-L4928) of the instance which is also used to populate the `hostname` field within the metadata. If you remove the cloud provider from the situation, we fall back to the OS hostname which may not match.
- The node name preference in kubelet, highest to lowest is: `hostname-override` > cloud provider `GetCurrentNodeName` > OS Hostname
- This PR maintains the existing node name in a backwards compatible way, and as such, should not affect any existing installations.

File: pkg/bootstrapper/cloud/cloud.go
Comment (Fedosin) [reply pkg/bootstrapper/cloud/cloud.go]: Done

**Comment by Fedosin** [comment]
"local-hostname" path is taken to comply with the MCO request: https://github.com/openshift/machine-config-operator/blob/master/templates/common/aws/files/usr-local-bin-aws-kubelet-nodename.yaml#L19

**Comment by JoelSpeed** [comment]
Perhaps we should replicate the comment from https://github.com/openshift/machine-config-operator/blob/b567492ff2b14444c805a3d8210768540d8a64eb/templates/common/aws/files/usr-local-bin-aws-kubelet-nodename.yaml#L15-L18 to this code as well, and/or link to the MCO counterpart?

File: pkg/bootstrapper/bootstrapper.go
Comment (sebsoto) [inline pkg/bootstrapper/bootstrapper.go:573]: After seeing this, I am wondering if it would make more sense to introduce a cloud interface
with `GetKubeletHostnameOverride` as method

My original and current thoughts are that it would need to be done only if there were more cloud specific functions to be introduced. Is this the only cloud specific WMCB function required for the CCM work @Fedosin?

I could go either way, I will let others weigh in, if they have an opinion on this.

**Review by sebsoto — CHANGES_REQUESTED** [review]
Thanks @Fedosin 
a few requests/questions

File: Makefile
Comment (Fedosin) [reply Makefile:14]: I think the problem here is that we `cannot use -c flag with multiple packages`. I.e. we can build a binary for just one package.
If we add unit tests for the "cloud" package, another binary has to be built.

File: test/e2e/bootstrapper_test.go
Comment (Fedosin) [reply test/e2e/bootstrapper_test.go:65]: Oh, I got it, you run these test on AWS, so metadata service will be available. In this case, I changed the platform type to "aws" in the tests

File: pkg/cloud/cloud.go
Comment (Fedosin) [reply pkg/cloud/cloud.go:17]: Done

File: pkg/bootstrapper/bootstrapper.go
Comment (Fedosin) [reply pkg/bootstrapper/bootstrapper.go:573]: Yeah, nothing else besides that is required. So, I don't think this interface is necessary.

File: pkg/cloud/cloud.go
Comment (JoelSpeed) [inline pkg/cloud/cloud.go:23]: Why not just return here, then you don't need the final return or to have the vars on 15 and 16. I think this would make it cleaner
```suggestion
		return getAWSMetadataHostname()
```

File: test/e2e/bootstrapper_test.go
Comment (JoelSpeed) [inline test/e2e/bootstrapper_test.go:65]: Why not use the constant `configv1.AWSPlatformType` for this instead of a raw string?

Comment (Fedosin) [reply test/e2e/bootstrapper_test.go:65]: we need to add openshift/api in the dependencies then...

Comment (Fedosin) [reply test/e2e/bootstrapper_test.go:65]: with all its dependencies it is 110k lines of code (146 files changed, 109510 insertions(+), 553 deletions(-)) just for a string constant.

Comment (Fedosin) [reply test/e2e/bootstrapper_test.go:65]: that's why I prefer not to do this

File: Makefile
Comment (aravindhp) [reply Makefile:14]: Is there any option that allows us to build a unit test binary that runs through all the packages?

File: pkg/bootstrapper/bootstrapper.go
Comment (aravindhp) [inline pkg/bootstrapper/bootstrapper.go:25]: Unnecessary new line. This will prevent the formatter from organizing this correctly.

File: pkg/cloud/cloud.go
Comment (aravindhp) [reply pkg/cloud/cloud.go:17]: This is a bit strange. It makes it seem that the only supported platfrom is AWS. Should we think of `platformType` to be a required parameter? Internally it can be a no-op for certain platforms.

File: pkg/bootstrapper/bootstrapper.go
Comment (sebsoto) [reply pkg/bootstrapper/bootstrapper.go:25]: This is compliant with the regular formatting we do for imports

first group built in, second external imports, third imports from other packages within the module.

Comment (aravindhp) [reply pkg/bootstrapper/bootstrapper.go:25]: Ah yes. Sorry my bad. Please ignore.

File: Makefile
Comment (Fedosin) [reply Makefile:14]: I don't think so, here they explain why it's not possible: https://github.com/golang/go/issues/15513#issuecomment-216694348

**Comment by aravindhp** [comment]
/retitle OCPCLOUD-1247: Set correct kubelet hostname for AWS platform

To keep the current patterns we have in the repo. I also suggest prefixing the commits with `[vendor]` and `[bootstrapper]`. We try not to use Jira prefixes for the individual commits.

**Review by JoelSpeed — COMMENTED** [review]
/lgtm

File: pkg/cloud/cloud.go
Comment (sebsoto) [reply pkg/cloud/cloud.go:17]: @Fedosin
@aravindhp made a good point here, the default case should return no error, as we should be able to pass any supported cloud platform from WMCO without error. Sorry for the misdirection :pray:

**Review by sebsoto — CHANGES_REQUESTED** [review]
One last comment, will approve after its change

File: pkg/cloud/cloud.go
Comment (Fedosin) [reply pkg/cloud/cloud.go:16]: Ack! It's fixed now.

**Comment by sebsoto** [comment]
/approve
Thank you for your work @Fedosin !

**Comment by sebsoto** [comment]
/hold
to ensure ccm test passes

**Pull request merged**
